File: backend/src/main/java/com/rikai/backend/model/InternshipBatch.java
package com.rikai.backend.model;

import com.rikai.backend.model.Enum.BatchStatus;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@FieldDefaults(level = AccessLevel.PRIVATE)
@Table(name = "internship_batches")
public class InternshipBatch {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Long id;

    @Column(name = "name", nullable = false)
    String name;

    @Column(name = "start_date", nullable = false)
    LocalDate startDate;

    @Column(name = "end_date", nullable = false)
    LocalDate endDate;

    @Column(name = "description", columnDefinition = "TEXT")
    String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20)
    @Builder.Default
    BatchStatus status = BatchStatus.ONGOING;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    Instant createdAt;

    @OneToMany(mappedBy = "internshipBatch", cascade = CascadeType.ALL)
    List<Intern> interns;
}

File: frontend/src/types/internshipBatch.js
/**
 * @enum {string}
 */
export const BatchStatus = {
    ACTIVE: 'ACTIVE',
    INACTIVE: 'INACTIVE',
    COMPLETED: 'COMPLETED'
}

/**
 * @typedef {Object} InternshipBatch
 * @property {number} id
 * @property {string} name
 * @property {string} startDate
 * @property {string} endDate
 * @property {string} [description]
 * @property {keyof typeof BatchStatus} [status]
 * @property {string} [createdAt]
 * @property {import('./intern').Intern[]} [internResponses]
 */

/**
 * @typedef {Object} InternshipBatchCreationRequest
 * @property {string} name
 * @property {string} startDate
 * @property {string} endDate
 * @property {string} [description]
 */

/**
 * @typedef {Object} InternshipBatchUpdateRequest
 * @property {string} [name]
 * @property {string} [startDate]
 * @property {string} [endDate]
 * @property {string} [description]
 * @property {keyof typeof BatchStatus} [batchStatus]
 * @property {string} [picMentorId]
 */

/**
 * @returns {InternshipBatchCreationRequest}
 */
export const createInternshipBatchCreationRequest = () => ({
    name: "",
    startDate: "",
    endDate: "",
    description: ""
})

/**
 * @returns {InternshipBatchUpdateRequest}
 */
export const createInternshipBatchUpdateRequest = () => ({
    name: "",
    startDate: "",
    endDate: "",
    description: "",
    batchStatus: null,
    picMentorId: ""
})
